### PR TITLE
Vca reward

### PR DESCRIPTION
**Pull request by Zeegomo**

### CODE REVIEW

File: src/rewards/mod.rs [hunk 1]
```diff
@@ -1,3 +1,8 @@
 pub mod community_advisors;
 pub mod veterans;
 pub mod voters;
+
+use rust_decimal::Decimal;
+pub type Funds = Decimal;
+// Lets match to the same type as the funds, but naming it funds would be confusing
+pub type Rewards = Funds;
```
Comment (Mr-Leshiy) [inline src/rewards/mod.rs:8]: Maybe better in this case to alias to the `Decimal` type explicitly

File: src/rewards/veterans.rs [hunk 2]
```diff
@@ -1,32 +1,326 @@
...
+    rewards_thresholds: EligibilityThresholds,
+    reputation_thresholds: EligibilityThresholds,
+) -> HashMap<VeteranAdvisorId, VeteranAdvisorIncentive> {
+    let final_rankings_per_review = veteran_rankings
+        .iter()
+        .into_group_map_by(|ranking| ranking.review_id())
+        .into_iter()
+        .map(|(review, rankings)| (review, calc_final_ranking_per_review(&rankings)))
+        .collect::<BTreeMap<_, _>>();
+    dbg!(&final_rankings_per_review);
```
Comment (Mr-Leshiy) [inline src/rewards/veterans.rs:116]:
```suggestion
```

File: src/rewards/veterans.rs [hunk 3]
```diff
@@ -1,32 +1,326 @@
...
+        .filter(|ranking| {
+            final_rankings_per_review
+                .get(&ranking.review_id())
+                .unwrap()
+                .is_positive()
+                == ranking.score().is_positive()
+        })
+        .counts_by(|ranking| ranking.vca.clone());
+
+    dbg!(&rankings_per_vca);
```
Comment (Mr-Leshiy) [inline src/rewards/veterans.rs:133]:
```suggestion
```

File: src/rewards/veterans.rs [hunk 4]
```diff
@@ -1,32 +1,326 @@
...
+    dbg!(&eligible_rankings_per_vca);
```
Comment (Mr-Leshiy) [inline src/rewards/veterans.rs:134]:
```suggestion
```

**Review by Mr-Leshiy — APPROVED** [review]
LGTM

File: src/rewards/veterans.rs [hunk 5]
```diff
@@ -1,32 +1,322 @@
...
+        self.advisor_review_row.score()
+    }
+
+    fn review_id(&self) -> AdvisorReviewId {
+        self.advisor_review_row.id()
+    }
+}
+
+fn calc_final_ranking_per_review(rankings: &[impl Borrow<VeteranRankingRow>]) -> ReviewRanking {
+    let rankings_majority = rankings.len() / 2;
```
Comment (eugene-babichenko) [inline src/rewards/veterans.rs:46]: What is the type of `rankings_majority`? Is it going to lead to any possible mistakes? I'm always a bit wary of divisions. Maybe employ decimals here as well?

File: src/rewards/veterans.rs [hunk 6]
```diff
@@ -1,32 +1,322 @@
...
-pub type VeteranAdvisorReward = rust_decimal::Decimal;
 
-pub type VeteranAdvisorRewards = Vec<(VeteranAdvisorId, VeteranAdvisorReward)>;
+#[derive(Serialize)]
+pub struct VeteranAdvisorIncentive {
+    pub rewards: Rewards,
+    pub reputation: u64,
+}
+
+pub type EligibilityThresholds = std::ops::RangeInclusive<usize>;
```
[the remark this answers is not in the document]
Comment (Zeegomo) [reply src/rewards/veterans.rs:21]: yeah, it's a bit counterintuitive the way it's written, but it was to keep naming close to the requirements

**Comment by Zeegomo** [comment]
I had to refactor `VeteranRankingRow` as `#[serde(flatten)]` was not behaving as expect

**Review by eugene-babichenko — APPROVED** [review]
The code in general is LGTM, let's see how it survives real-life testing.

**Pull request merged**